spec/abci/abci++_app_requirements.md

```diff
@@ -188,6 +192,34 @@ Likewise, `ExtendVote` can also be non-deterministic:
 * *w<sup>r</sup><sub>p</sub> = w<sup>r</sup><sub>q</sub> &#8655;
   e<sup>r</sup><sub>p</sub> = e<sup>r</sup><sub>q</sub>*
 
+### Mempool Connection Requirements
+
+Before defining our notation, we need to introduce a requirement for `CheckTx` to always
+yield the same result while *p*'s Application is at a given height.
+
+* Requirement 13 [`CheckTx`, same-height immutability]: For any correct process *p*,
+  any transaction *tx*, and any height *h<sub>p</sub>*,
+  `ResponseCheckTx` returns the same result code while *p* is at *h<sub>p</sub>*.
+
+Requirement 13 allows us to use heights rather than time in our notation.
+Let *CheckTxCode<sub>tx,p,h</sub>* denote the result  code returned by *p*'s Application, via `ResponseCheckTx`,
+to CometBFT's `RequestCheckTx` call occurring while the Application is at height *h*
+and having transaction *tx* as parameter.
+Let predicate *OK(CheckTxCode)* denote whether *CheckTxCode* is `SUCCESS`.
+
+* Requirement 14 [`CheckTx`, eventual non-oscillation]: For any correct process *p*
+  and any transaction *tx*, there exists a height *h<sub>p,stable</sub>*
+  such that *OK(CheckTxCode<sub>tx,p,h<sub>p,stable</sub></sub>) = OK(CheckTxCode<sub>tx,p,h</sub>)*
+  for any height *h > h<sub>p,stable</sub>*.
+
+Requirement 14, local to process *p*, ensures that
+a transaction will eventually stop oscillating between `CheckTx` success and failure
+if it stays in *p's* mempool for long enough.
+Thus, whilst being a requirement local to *p*, it ensures that
+a transaction will leave the mempool of *all* full nodes,
+either because it is expunged everywhere due to failing `CheckTx` calls,
+or because it stays valid long enough to be gossipped, proposed and decided.
```

Comment:
Echoing Hernán above, if the transaction is decided an application, by implementing replay protection, will start rejecting the same transaction.

Reply:
Yes, but that' beyond (my understanding) of the requirement to the app. The requirements should tell the app _how_ to do stuff, but rather _what_ they can and cannot do.

A similar example is the requirement on PrepProp and ProcProp, we don't says "you should of shouldn't used the past state for your checks"... that's up to the app devs to decide. We just say what the "limits" are to what they can and cannot do. To me replay protection is a similar thing. It's up to the app to decided whether, or how to implement it. We just need to set the limits, so that the mempool functions correctly.